packages/pigeon/lib/generator.dart

```diff
@@ -0,0 +1,15 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'ast.dart';
+import 'generator_tools.dart';
+
+/// A superclass of generator classes.
+///
+/// This provides the structure that is common across generators for different languages.
+abstract class Generator<T> {
+  /// Generates files for specified language with specified [languageOptions]
+  void generate(
+      T languageOptions, Root root, StringSink sink, FileType fileType);
```

Comment:
Thanks for making the adjustment.  I think having a Generator interface and maintaining the Adapter keeps the generator files simple.  Everything looks good but one thing.  I think I missed some discussion on `FileType`.  I don't think you've really gained anything by its inclusion.  I think it's a design mistake for the following reasons:

1) Implementing an abstract function and asserting that one parameter is exactly one value means the abstraction doesn't match the implementations well.
1) The `FileType` already has 2 omissions: The AST Generator and the Dart Test generator whose output is not a "source" or a "header"
1) Considering input to some generators being "header" doesn't make sense in many of the generators.
1) The design isn't modular in that new types of output has to be added to a finite list of outputs and all Generators would have to consider what it means to them.

What do we think we gained by adding them?  Now there is a grouping of similar generators: CppHeader and CppSource.  That seems like it complicates the design with little benefit.  What do you think?

Reply:
Thanks for the feedback.

Would the inclusion of a `default` option solve some of these issues?

This allows for generators that don't have multiple file types to ignore the `fileType` parameter without mislabeling them. It would also avoid the need to add more options to the FileType enum in the future (unless there is some other multi file system that isn't covered by `source` and `header`).

Is there a better way to handle combining the generator classes without the use of a parameter similar to this implementation?

Reply:
The Generator abstraction is already powerful enough to capture the idea of a group.

![VP9DJiCm48NtSufHLxJW2qHHQ8LGMI21b5p0YeTYDTYMnqP2qBlZ_BGDeUacSjvxy_IDvJK7wKFyQ02q2UJVP4su9KDU1klpgi1lGBp5NI_HZNL1MyCPLdSeEMIuATE9jTbdMARl4WuNgpPAkYGETVnkOwEsnz9bFtlVUP-oqQfLZ_N9VTOSyaVlhIzcS5xrZgnvwN-leFq3MWbnfWIe6ycC3ywTj8HyE8zX_fC6nZxy2Qzw](https://user-images.githubusercontent.com/30870216/210428942-961a1c53-19b5-41e3-9a9a-ba1e4cd91212.png)

or

![VP1D3e8m44RtSug9Aq6v02482yF6H1EuG4ChDjQMjCMDUdSB5I7nPzDCt_Hx-TBCMA9jTn40N5gcZwHcM339DB5A9rMADq1SOUCHMwhMSYLDZDKQYR4nvgMR39Vd64jt1l3ugiefQHrywSn9TO8MepJmsSsmknB1QKz7lTlkkB79Lckbqnzr3hnXIkxzzK-rZq9X54qj0Mf1Z9b0eLNNkjBpf6TzRX4kPjdtXCbzBXtU0sHC](https://user-images.githubusercontent.com/30870216/210428970-a6c1835c-ffe1-4367-9f75-40ed29504dcf.png)

I still don't think that the group is buying you much in terms of making the codebase easier to understand or work with though.

Reply:
> What do we think we gained by adding them? Now there is a grouping of similar generators: CppHeader and CppSource.

The issue is that they aren't "similar" they are two fundamentally linked things that can't meaningfully operate on their own. To be useful, each one *must* be paired with a call to the other with identical inputs. And changes to the header generator almost always require exactly corresponding changes to the source generator or everything breaks.

That's why I would like to see combined classes for all of the linked groups; the goal here is to move all of the generator logic into the subclasses, and inherently codependent logic should be in the same class.

(Having two parallel classes, each with half the logic, feels like when someone has two Ivar arrays where the object indexes must match, instead of using a map or an array of structures. Technically it works, but it's really fragile and harder to understand.)

Reply:
Working on a design doc for this to allow for more effective communication about our goals here. https://docs.google.com/document/d/1ABr5yZM_sOZTd66bIVpsCJ7_koKmH0tZuLH4ZT1cs8c/edit?usp=sharing